wordpress-mobile/WordPress-Android
# Move logic from photo picker fragment to view model

This PR completely moves the logic out of the PhotoPickerFragment to PhotoPickerViewModel. I was trying to separate it into more PRs but it got complicated. I think the best way to review this PR is by going step by step through the test scenarios. There is a ui model for the action mode which is now pulled out of the fragment. Let me know what you think about this solution. Everything related to the ViewState should be in the UI state data class.

I'll add more unit tests in another PR. I didn't want to make this one massive

To test 1 - Permissions (try comparing to the play store version):
- Start with a clean version of the app
- Login 
- Go to "Add new post" with Gutenberg
- Add an Image block
- Click on "Choose from device"
- Notice that the photo picker opens but it asks for permissions to access storage
- Notice that the FAB is not visible
- Notice that the bottom bar is not visible
- Permanently deny the permissions 
- Go back to the same screen
- Notice that the message is different and the link is to the settings to enable the permissions
- Enable the permissions
- You see the media list and the bottom bar

To test 2 - Permissions with Gallery block
- The same as test 1, only add a Gallery block

To test 3 - Permissions with Video block
- The same as test 1, only add a Video block

To test 4 - Permissions with Gallery block
- Start with a clean version of the app
- Login 
- Go to "Add new post"
- Switch to classic editor
- Click on the "+" button to add media
- Follow the permissions flow from test 1

To test 5 - Permissions with Stories
- Start with a clean version of the app
- Login 
- Go to "Add a story"
- Notice that you see the permissions screen to access storage
- Notice that you see the the FAB camera button
- The rest of the behaviour should be the same as in the test 1 (compare with the play store version)

To test 6 - Image block
- With enabled permissions follow steps 1-5 from test 1
- Notice that the media bottom bar is shown (with an image icon)
- Click on an image 
- Notice that the Action mode is started
- Notice the title
- Notice that there is no Check button in the action mode toolbar
- Notice that the bottom bar changes to "Edit"/"Insert" bottom bar
- Rotate your device
- Notice that everything stays the same
- Click on "Back" button in toolbar
- Notice that the selection disappears
- Notice that the bottom bar changes to "Media" 
- Notice that the action mode disappears

To test 7 - Video block
- Follow steps from test 6 with a video block

To test 8 - Gallery block
- With enabled permissions follow steps 1-3 from test 1 - select a gallery block
- Notice that the media bottom bar is shown (with an image icon)
- Click on an image 
- Notice the action mode is started
- Notice the title is "1 selected" 
- Notice there is the "Edit"/"Insert" bottom bar
- Notice the item is selected with a number 1 next to it
- Click on another image
- Notice the title is "2 selected" 
- Notice the second item has number 2 next to it
- Click on the first image
- Notice that the second image has number 1 next to it
- Try rotating the device
- Everything should work as expected

To test 9 - Stories
- Add a story post
- Notice the bottom bar has 2 options - choose from WP library/the device
- Click on an item
- Notice that the bottom bar disappears
- Notice that there is a "Check" button in the toolbar
- Click on the same item
- Notice that it gets unselected and the action mode disappears
- Add a few more items
- Click on the check button 
- See that all the items are added to the story 

To test 10 - Aztec editor
- Add a post
- Switch to classic editor
- Click on the "+" button
- Notice a list of images and videos are shown
- Click on a few
- Notice an action mode with a "Check" icon appears
- Unselect everything 
- The action mode disappears
- Select an item and click on "Check" 
- The item gets added to the post
 
PR submission checklist:
